django/middleware/csrf.py

```diff
@@ -36,8 +39,38 @@ def _get_failure_view():
     return get_callable(settings.CSRF_FAILURE_VIEW)
 
 
-def _get_new_csrf_key():
-    return get_random_string(CSRF_KEY_LENGTH)
+def _get_new_csrf_nonce():
+    return get_random_string(CSRF_NONCE_LENGTH, allowed_chars=CSRF_ALLOWED_CHARS)
+
+
+def _pad_cipher_nonce(nonce):
+    """
+    Given a nonce (assumed to be a string of CSRF_ALLOWED_CHARS), generate a
+    token by adding a pad and using it to encrypt the nonce.
+    """
+    pad = _get_new_csrf_nonce()
+    chars = CSRF_ALLOWED_CHARS
+    pairs = zip([chars.index(x) for x in nonce], [chars.index(x) for x in pad])
```

Comment:
I think this could use a generator expression instead of list comprehension to avoid additional temporary lists in memory. For example:

```
pairs = zip((chars.index(x) for x in nonce), (chars.index(x) for x in pad))
```

(Same in `_unpad_cipher_token()`)